// @camunda8/sdk/dist/c8/lib/C8Dto.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PublishMessageResponse = exports.CorrelateMessageResponse = exports.BroadcastSignalResponse = exports.CreateProcessInstanceResponse = exports.DeployResourceResponse = exports.DeployResourceResponseDto = exports.FormDeployment = exports.DecisionRequirementsDeployment = exports.DecisionDeployment = exports.ProcessDeployment = exports.RestApiJob = void 0;
const lib_1 = require("../../lib");
class RestApiJob extends lib_1.LosslessDto {
}
exports.RestApiJob = RestApiJob;
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "jobKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "elementInstanceKey", void 0);
__decorate([
    lib_1.Int64String
], RestApiJob.prototype, "deadline", void 0);
class ProcessDeployment extends lib_1.LosslessDto {
}
exports.ProcessDeployment = ProcessDeployment;
__decorate([
    lib_1.Int64String
], ProcessDeployment.prototype, "processDefinitionKey", void 0);
class DecisionDeployment extends lib_1.LosslessDto {
}
exports.DecisionDeployment = DecisionDeployment;
__decorate([
    lib_1.Int64String
], DecisionDeployment.prototype, "decisionKey", void 0);
__decorate([
    lib_1.Int64String
], DecisionDeployment.prototype, "dmnDecisionRequirementsKey", void 0);
class DecisionRequirementsDeployment extends lib_1.LosslessDto {
}
exports.DecisionRequirementsDeployment = DecisionRequirementsDeployment;
__decorate([
    lib_1.Int64String
], DecisionRequirementsDeployment.prototype, "dmnDecisionRequirementsKey", void 0);
class FormDeployment {
}
exports.FormDeployment = FormDeployment;
__decorate([
    lib_1.Int64String
], FormDeployment.prototype, "formKey", void 0);
class DeployResourceResponseDto extends lib_1.LosslessDto {
}
exports.DeployResourceResponseDto = DeployResourceResponseDto;
__decorate([
    lib_1.Int64String
], DeployResourceResponseDto.prototype, "deploymentKey", void 0);
class DeployResourceResponse extends DeployResourceResponseDto {
}
exports.DeployResourceResponse = DeployResourceResponse;
class CreateProcessInstanceResponse {
}
exports.CreateProcessInstanceResponse = CreateProcessInstanceResponse;
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processDefinitionKey", void 0);
__decorate([
    lib_1.Int64String
], CreateProcessInstanceResponse.prototype, "processInstanceKey", void 0);
/** The signal was broadcast. */
class BroadcastSignalResponse extends lib_1.LosslessDto {
}
exports.BroadcastSignalResponse = BroadcastSignalResponse;
__decorate([
    lib_1.Int64String
    /** The unique ID of the signal that was broadcast. */
], BroadcastSignalResponse.prototype, "signalKey", void 0);
class CorrelateMessageResponse extends lib_1.LosslessDto {
}
exports.CorrelateMessageResponse = CorrelateMessageResponse;
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "key", void 0);
__decorate([
    lib_1.Int64String
], CorrelateMessageResponse.prototype, "processInstanceKey", void 0);
class PublishMessageResponse extends lib_1.LosslessDto {
}
exports.PublishMessageResponse = PublishMessageResponse;
__decorate([
    lib_1.Int64String
], PublishMessageResponse.prototype, "key", void 0);
